Add calving law tests for Humboldt test group

matthewhoffman opened the pull request:
This merge introduces decomposition and restart tests to the Humboldt test group.  It also adds the ability to test them using different velocity solvers and calving laws.  It then creates test suites for the various calving law tests.  There are enough of these that are expensive enough that we don't want to add them all to our standard regression suites.  The new suites allow bulk testing of calving options that can be run periodically when changes to the calving code  occur.  Note that not all of these tests pass due to known (or now exposed) issues in MALI.  Indeed, a primary purpose of this PR is to expose these sort of bugs.  Failures in the calving integration suites will be resolved in MALI in the future.

Additionally, a set of "full physics" configurations are created that also include damage threshold calving and facemelting.  These configurations are meant to represent the fullest set of physics we currently use in MALI for a realistic mesh.  These full physics tests are added to he calving regression suites, as well as the standard integration suites.

pep8speaks commented:
Hello @matthewhoffman! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`compass/landice/tests/humboldt/__init__.py`](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/__init__.py):

> [Line 22:18](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/__init__.py#L22): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 22:28](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/__init__.py#L22): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 24:33](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/__init__.py#L24): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 36:41](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/__init__.py#L36): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 39:69](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/__init__.py#L39): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`compass/landice/tests/humboldt/decomposition_test/__init__.py`](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/decomposition_test/__init__.py):

> [Line 3:1](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/decomposition_test/__init__.py#L3): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 9:80](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/decomposition_test/__init__.py#L9): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 17:1](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/decomposition_test/__init__.py#L17): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 50:12](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/decomposition_test/__init__.py#L50): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 52:12](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/decomposition_test/__init__.py#L52): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 61:9](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/decomposition_test/__init__.py#L61): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

* In the file [`compass/landice/tests/humboldt/restart_test/__init__.py`](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/restart_test/__init__.py):

> [Line 43:12](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/restart_test/__init__.py#L43): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 45:12](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/restart_test/__init__.py#L45): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four

* In the file [`compass/landice/tests/humboldt/run_model.py`](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py):

> [Line 42:58](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py#L42): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 82:12](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py#L82): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 84:12](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py#L84): [E111](https://duckduckgo.com/?q=pep8%20E111) indentation is not a multiple of four
> [Line 85:72](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py#L85): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 86:23](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py#L86): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 87:23](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py#L87): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 104:40](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py#L104): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 114:9](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py#L114): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 116:9](https://github.com/MPAS-Dev/compass/blob/caec9b3fdd78ce236f075ebc0098558ed674b7fc/compass/landice/tests/humboldt/run_model.py#L116): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

xylar commented:
@matthewhoffman, whenever it's convenient (perhaps when you're doing the PEP8 fixes), could you rebase onto `master` now that #320 is in?  I get an extra email every time CI fails so it would be nice to have it working again.

matthewhoffman commented:
Sure, I will do that.  You get an email every time any PRs fail CI?  Is that only triggered when a new PR is opened or an existing PR is modified (e.g. new commits are pushed)?  Or does it happen nightly on open PRs?

xylar commented:
I get an email every time you push a new commit to an open PR and CI fails.  So that's a lot of emails.  They're easy to throw away but there's a chance something actually important gets lost.

trhille commented:
@matthewhoffman, note that the eigencalving tests fail on execution due to the >10% failed ablation error. I reduced `config_calving_eigencalving_parameter_scalar_value` from 3.14e16 to 3.14e14 and got rid of that error.

matthewhoffman commented:
I fixed PEP8 issues and rebased this branch so the spack build of Albany within COMPASS is available.  I retested the two test suites in this PR and got these results:

`humboldt_calving_tests`:
```
Test Runtimes:
00:48 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-none
00:47 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-floating
00:06 FAIL landice_humboldt_mesh-default_decomposition_test_velo-none_calving-eigencalving
00:47 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-specified_calving_velocity
00:46 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-von_mises_stress
00:46 PASS landice_humboldt_mesh-default_decomposition_test_velo-none_calving-damagecalving
00:48 FAIL landice_humboldt_mesh-default_decomposition_test_velo-none_calving-ismip6_retreat
00:19 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-none
00:19 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-floating
00:04 FAIL landice_humboldt_mesh-default_restart_test_velo-none_calving-eigencalving
00:19 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-specified_calving_velocity
00:19 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-von_mises_stress
00:20 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-damagecalving
00:20 PASS landice_humboldt_mesh-default_restart_test_velo-none_calving-ismip6_retreat
Total runtime 06:52
FAIL: 3 tests failed, see above.
```

Note that the goal for this PR is not to have all tests pass, but to provide tests for each configuration.

`humboldt_calving_tests_FO` results will be posted in a later comment.

matthewhoffman commented:
The runs with FO solver on this 1km Humboldt mesh are (not surprisingly) very slow.  For the sake of the calving tests, the 3km Humboldt mesh would probably be sufficient and would be much faster.  @trhille , do you think that makes sense?  If so, can you provide a 3km Humboldt mesh to put on Anvil?  (and let me know the yaml settings to use). I can update the test suite accordingly.

matthewhoffman commented:
After switching to 3km mesh, same tests pass and fail, but suite runs twice as fast:
```
Test Runtimes:
00:11 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-none
00:11 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-floating
00:02 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-eigencalving
00:11 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-specified_calving_velocity
00:10 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-von_mises_stress
00:11 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-damagecalving
00:12 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-ismip6_retreat
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-none
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-floating
00:04 FAIL landice_humboldt_mesh-3km_restart_test_velo-none_calving-eigencalving
00:19 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-specified_calving_velocity
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-von_mises_stress
00:19 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-damagecalving
00:19 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-ismip6_retreat
Total runtime 03:05
FAIL: 3 tests failed, see above.
```

matthewhoffman commented:
I adjusted the tolerances for the FO decomp tests to account for non-BFB decomp behavior of the FO solver.  After that I get the following results for the humboldt_calving_tests_FO suite:
```
Test Runtimes:
01:29 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-none
01:08 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-floating
01:19 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-eigencalving
01:13 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-specified_calving_velocity
01:25 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-von_mises_stress
01:14 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-damagecalving
01:41 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-ismip6_retreat
01:37 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-none
01:14 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-floating
01:21 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-eigencalving
01:18 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-specified_calving_velocity
01:27 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-von_mises_stress
01:17 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-damagecalving
01:39 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-ismip6_retreat
Total runtime 19:24
```

The one failure is a known issue that will be resolved later in MALI.

matthewhoffman commented:
This PR is now complete and ready for review.  I suggest running the `humboldt_calving_tests` suite (takes about 4 minutes on Badger) and the `humboldt_calving_tests_FO` suite (takes about 25 min.) and confirming you reproduce the results below.  (@xylar , you could skip the 2nd longer one and leave that for @trhille.)

`humboldt_calving_tests`:
```
Test Runtimes:
00:16 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-none
00:12 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-floating
00:14 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-eigencalving
00:13 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-specified_calving_velocity
00:12 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-von_mises_stress
00:12 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-damagecalving
00:13 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-ismip6_retreat
00:12 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-von_mises_stress_damage-threshold_faceMelting
00:19 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-none
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-floating
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-eigencalving
00:19 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-specified_calving_velocity
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-von_mises_stress
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-damagecalving
00:19 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-ismip6_retreat
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-von_mises_stress_damage-threshold_faceMelting
Total runtime 04:13
FAIL: 2 tests failed, see above.
```

`humboldt_calving_tests_FO`:
```
Test Runtimes:
01:29 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-none
01:23 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-floating
01:26 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-eigencalving
01:34 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-specified_calving_velocity
01:32 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-von_mises_stress
01:18 PASS landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-damagecalving
01:50 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-ismip6_retreat
02:00 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-von_mises_stress_damage-threshold_faceMelting
01:36 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-none
01:30 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-floating
01:10 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-eigencalving
01:28 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-specified_calving_velocity
01:17 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-von_mises_stress
01:11 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-damagecalving
01:32 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-ismip6_retreat
00:54 PASS landice_humboldt_mesh-3km_restart_test_velo-fo_calving-von_mises_stress_damage-threshold_faceMelting
Total runtime 23:13
FAIL: 2 tests failed, see above.

```
 
I also tested the `full_integration` suite  - because a new test has been added - and confirmed all tests pass (with the exception of an expected BASELINE error for the new test).  No need for reviewers to necessarily run it - the newly added test to that suite is included in the above.

trhille commented:
My testing of the `humboldt_calving_tests` and `humboldt_calving_tests_FO` agree with @matthewhoffman's results above.

xylar commented:
Great, I'll test tomorrow.

trhille reviewed (commented):
@matthewhoffman, terrific work on this! It will be a huge help for future development. I have a handful on minor comments.

trhille reviewed (approved):
My comments are all essentially trivial, so I am approving this.

xylar commented on compass/landice/tests/humboldt/run_model.py:
```diff
@@ -0,0 +1,191 @@
+        velo_solver : {'sia', 'FO', 'none'}
+            The velocity solver setting to use for this test case
+
+        calving_law: {'none', 'floating', 'eigencalving',
+                      'specified_calving_velocity', 'von_Mises_stress',
+                      'damagecalving', 'ismip6_retreat'}, optional
+            The calving law setting to use for this test case. If not
+            specified, set to 'none'.
```
I wish this worked.  The docstring doesn't parse correctly:
![image](https://user-images.githubusercontent.com/4179064/164713559-00e3bd65-84dc-4df0-856f-493084bbef18.png)
Instead, I'd suggest:
```suggestion
        calving_law: str, optional
            The calving law setting to use for this test case. If not
            specified, set to 'none'. Valid values are: 'none', 
            'floating', 'eigencalving', 'specified_calving_velocity', 
            'von_Mises_stress', 'damagecalving', 'ismip6_retreat'}
```

matthewhoffman replied:
Fixed in ecfafe4cdb998425c88b87d97bb4ec84b8a53208

xylar commented on docs/developers_guide/landice/api.rst:
```diff
@@ -149,6 +149,31 @@ greenland
    run_model.RunModel.setup
    run_model.RunModel.run
 
+humboldt
+~~~~~~~~
```
Thanks for alphabetizing!

xylar reviewed (commented):
I'm seeing 3 test failures for each suite on Anvil with mvapich:
```
00:06 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-von_mises_stress
00:05 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-ismip6_retreat
00:05 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-von_mises_stress_damage-threshold_faceMelting
...
01:07 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-von_mises_stress
01:19 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-ismip6_retreat
01:01 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-von_mises_stress_damage-threshold_faceMelting
```
Results are in:
```
/lcrc/group/e3sm/ac.xylar/compass_1.0/anvil/test_20220422/mali_calving
/lcrc/group/e3sm/ac.xylar/compass_1.0/anvil/test_20220422/mali_calving_fo
```
if you want to investigate.

xylar commented:
@matthewhoffman, nope, it's fine to leave f-strings and all that.  I may go through the code and move to f-strings in a PR someday but there are plenty of places not using them now.  The comma thing is definitely no problem, I just mentioned it because it's -- unusual?

In terms of the test naming convention, are you happy with what you have?  if so I'll approve.

matthewhoffman commented:
Thanks, @xylar.  I actually never knew about f-strings until now, I think because I/we had been using python 2.7 for so long.  The comma thing is indeed a habit I picked up in early python days from tuples where such a comma is required.  I learned that trailing commas are allowed in other list-like data structures, and actually in some situations recommended (e.g. long lists with one entry on each line so that the formatting of the final item doesn't look different and items can be moved around easily).  My usage was not for some clever reason like that, just ignorance. :) Regarding test names, I'm going to discuss with Trevor when he gets here tomorrow.  There were good reasons for my strange naming scheme, but it's possible we can find something less unusual.

matthewhoffman commented on compass/landice/tests/humboldt/__init__.py:
```diff
@@ -14,4 +17,61 @@ def __init__(self, mpas_core):
+
+        # Set up tests without calving using the 1km mesh
+        mesh_type = '1km'
+        for velo_solver in ['FO', ]:
```
@trhille , I was trying to keep the number of new tests added by this PR to a minimum, so I deliberately only included the FO solver for the 1km mesh, since none of the 1km tests are in any suites.  My thought was it would be nice to be able to set up a high res Humboldt test with COMPASS in a standardized way.  I'm open to adding other versions of this test if you think they'd be useful.

matthewhoffman commented on compass/landice/tests/humboldt/decomposition_test/__init__.py:
```diff
@@ -0,0 +1,183 @@
+
+        # Commented code to make use of mesh generation step
+        # Note it will not include uReconstructX/Y or muFriction!
+        # It will also add a few minutes run time to the test!
```
@trhille , good point.  Given we don't expect the basal friction optimization to be automated any time soon, we probably will never completely set up the mesh this way.  I could imagine interpolating basal friction from an existing file just to be able to be able to test and end to end workflow, but that might be more trouble than it's worth.  I'll remove these comments.  If we ever want to do that, we can figure it out easily enough.

matthewhoffman replied:
Removed in 264047c2426e13dd99c4bed2be0ecffdb5021855

matthewhoffman commented on compass/landice/tests/humboldt/restart_test/__init__.py:
```diff
@@ -0,0 +1,135 @@
+        variables = ['thickness', 'surfaceSpeed']
+
+        if self.calving_law is not None and self.calving_law != 'none':
+            variables.append('calvingThickness')
```
That's a good suggestion.  We know most of the trouble is likely to come in the conversion of `calvingVelocity` to `calvingThickness`, so it is likely to be useful to check both fields.  Given these tests are already pretty "heavy" I don't have a problem with including an additional field verification.

matthewhoffman replied:
I added calvingVelocity in 74a03d326c44a2a26fbdcb0f07e46d5ef7dd6d3f.  This already came in handy for debugging the ismip6-retreat issues I worked on today.

matthewhoffman commented on compass/landice/tests/humboldt/albany_input.yaml:
```diff
@@ -0,0 +1,277 @@
+
+# Discretization Description
+  Discretization:
+    Exodus Output File Name: albany_output.exo
```
I commented that line in e8ef60ec4193568836b7f3b3f68f53136d082aa6.

matthewhoffman commented on compass/landice/tests/humboldt/decomposition_test/__init__.py:
```diff
@@ -0,0 +1,183 @@
+from compass.validate import compare_variables
+from compass.testcase import TestCase
+# from compass.landice.tests.humboldt.mesh import Mesh
```
Removed in 264047c2426e13dd99c4bed2be0ecffdb5021855

matthewhoffman commented on compass/landice/tests/humboldt/decomposition_test/__init__.py:
```diff
@@ -0,0 +1,183 @@
+    """
+    A test case for performing two MALI runs of a humboldt setup, one with one
+    core and one with four.  The test case verifies that the results of the
+    two runs are identical or close to identical.
```
Added in 264047c2426e13dd99c4bed2be0ecffdb5021855

matthewhoffman commented on compass/landice/tests/humboldt/restart_test/streams.restart.rst:
```diff
@@ -0,0 +1,15 @@
+
+<stream name="output"
+        output_interval="0000-00-01_00:00:00"
+        reference_time="0000-01-01_00:00:00"
```
Fixed in f45527f759e522ba1e140b900ace1623d1a5ebfc.  I also changed start dates and references dates to January 1.  That exposed some new failures that I've fixed in https://github.com/MALI-Dev/E3SM/pull/32 (but will appear as failures when running the tests in this PR with MALI-Dev/develop).  So it was useful to take a closer look at these settings.

matthewhoffman commented on docs/developers_guide/landice/test_groups/humboldt.rst:
```diff
@@ -23,11 +23,39 @@ The class :py:class:`compass.landice.tests.humboldt.mesh.Mesh`
+------------------
+The compass.landice.tests.humboldt.decomposition_test.DecompositionTest
+performs the same simulation on different numbers of cores. It ensures
+relevant variablesd are identical or have expected differences.
```
Fixed in bcd29cf9d3806d57deca00b3ba07b5ca95df2529

matthewhoffman commented on docs/developers_guide/landice/test_groups/humboldt.rst:
```diff
@@ -23,11 +23,39 @@ The class :py:class:`compass.landice.tests.humboldt.mesh.Mesh`
+restart_test
+------------
+The compass.landice.tests.humboldt.restart_test.RestartTest performs a
+run of a full specified duration followe by a short run plus a restart
```
Fixed in bcd29cf9d3806d57deca00b3ba07b5ca95df2529

matthewhoffman commented on compass/landice/suites/humboldt_calving_tests_FO.txt:
```diff
@@ -0,0 +1,18 @@
+landice/humboldt/mesh-3km_decomposition_test/velo-fo_calving-none
```
Changed in 7615549c37f49061a384444f27c2302c9f8386dd

matthewhoffman commented on compass/landice/suites/humboldt_calving_tests.txt:
```diff
@@ -0,0 +1,17 @@
+landice/humboldt/mesh-3km_decomposition_test/velo-none_calving-eigencalving
+landice/humboldt/mesh-3km_decomposition_test/velo-none_calving-specified_calving_velocity
+landice/humboldt/mesh-3km_decomposition_test/velo-none_calving-von_mises_stress
+landice/humboldt/mesh-3km_decomposition_test/velo-none_calving-damagecalving
```
For various reasons that option happens to be spelled `damagecalving` in MALI.

matthewhoffman commented on compass/landice/tests/humboldt/decomposition_test/__init__.py:
```diff
@@ -0,0 +1,183 @@
+        self.face_melt = face_melt
+
+        # build dir name.  always include velo solver and calving law
+        subdir = 'mesh-{}_decomposition_test/velo-{}_calving-{}'.format(
+                 mesh_type, velo_solver.lower(), calving_law.lower())
+        # append damage and facemelt if provided
+        if damage is not None:
+            subdir += '_damage-{}'.format(damage)
+        if face_melt is True:
+            subdir += '_faceMelting'
```
I removed all the hanging commas, but left the `.format` strings.  I'll switch to using f-strings in the future.

matthewhoffman commented:
@xylar (and @trhille ), I've addressed all the issues in the reviews, with a few exceptions as discussed above.  I've done extensive retesting that revealed a few new issues that I either fixed in this PR or require MALI changes (https://github.com/MALI-Dev/E3SM/pull/32).  This PR is ready to be reviewed again as desired and then merged.  When you are happy with it, feel free to merge it or assign it to me or Trevor to merge.

For the record, here are the current test suite results:
`humboldt_calving_tests` suite:
```
Test Runtimes:
00:12 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-none
00:12 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-floating
00:12 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-eigencalving
00:12 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-specified_calving_velocity
00:12 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-von_mises_stress
00:12 PASS landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-damagecalving
00:12 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-ismip6_retreat
00:11 FAIL landice_humboldt_mesh-3km_decomposition_test_velo-none_calving-von_mises_stress_damage-threshold_faceMelting
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-none
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-floating
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-eigencalving
00:17 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-specified_calving_velocity
00:17 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-von_mises_stress
00:18 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-damagecalving
00:35 FAIL landice_humboldt_mesh-3km_restart_test_velo-none_calving-ismip6_retreat
00:17 PASS landice_humboldt_mesh-3km_restart_test_velo-none_calving-von_mises_stress_damage-threshold_faceMelting
Total runtime 04:14
FAIL: 3 tests failed, see above.
```
`humboldt_calving_tests_fo` suite:
```
Test Runtimes:
01:26 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-none
01:03 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-floating
01:14 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-eigencalving
01:08 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-specified_calving_velocity
01:17 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-von_mises_stress
01:08 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-damagecalving
01:32 ESC[91mFAILESC[0m landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-ismip6_retreat
01:12 ESC[91mFAILESC[0m landice_humboldt_mesh-3km_decomposition_test_velo-fo_calving-von_mises_stress_damage-threshold_faceMelting
00:56 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_restart_test_velo-fo_calving-none
00:49 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_restart_test_velo-fo_calving-floating
00:53 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_restart_test_velo-fo_calving-eigencalving
00:50 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_restart_test_velo-fo_calving-specified_calving_velocity
00:53 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_restart_test_velo-fo_calving-von_mises_stress
00:50 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_restart_test_velo-fo_calving-damagecalving
01:00 ESC[91mFAILESC[0m landice_humboldt_mesh-3km_restart_test_velo-fo_calving-ismip6_retreat
00:49 ESC[92mPASSESC[0m landice_humboldt_mesh-3km_restart_test_velo-fo_calving-von_mises_stress_damage-threshold_faceMelting
Total runtime 17:03
FAIL: 3 tests failed, see above.
```
Note that the 2 ismip6_retreat failures in both suites pass when using the MALI PR https://github.com/MALI-Dev/E3SM/pull/32, but that does not need to be tested here.  The one remaining failure in each suite we'll address at a later date.  

Regarding the Anvil failures you saw, I am currently without Anvil access as my primary computer is still being repaired and I haven't set up ssh keys on this other computer.  I've been given the impression my computer issue should be fixed any day now for over a week.  Anyway, I'd rather address those in a follow up PR if necessary, as getting this merged ASAP will help with some MALI development work we are working on.

xylar commented:
@matthewhoffman, it looks like a rebase is needed to handle a merge conflict with `docs/developers_guide/landice/api.rst`.  Once that's done, feel free to merge.  I have already assigned the PR to you awhile back.

matthewhoffman commented:
Ok, thanks, @xylar !  I saw you had approved it but I wanted to make sure you had the chance to look again if you wanted.  Do you prefer a rebase or can I fix the merge conflicts locally and push `master` to merge & close the PR?

xylar commented:
@matthewhoffman, either is fine.  I generally prefer a rebase so CI runs on exactly what you're going to get before you merge but in this case it's just docs.
